### PR TITLE
Some adjustments for KEYCLOAK-3056 / PR #3893

### PR DESCRIPTION
**Pull request by skjolber**
In addtion to fixing/adding some unit tests, an additional test was added for documents actually having a signature when wantAssertionsSigned is enabled.

### CODE REVIEW

**Comment by stianst** [comment]
Does this assume https://github.com/keycloak/keycloak/pull/3893 would be merged first? Or does it contain the changes?

Since you're both from Norway and are both interested in this is it safe to guess that you both work for the same company?

**Comment by stianst** [comment]
Never mind. I somehow missed that 3893 was already merged.

**Comment by skjolber** [comment]
I'll look into the unit test which are failing

**Comment by anderius** [comment]
@stianst: No, but we might be integrating with the same IdP. Sorry I never got to complete the other PR, as the priority fell when we realised the IdP signed the response as well as the assertion. Then I went for fathers leave...

**Comment by skjolber** [comment]
I am having some problems in recreating the failing unit tests locally, so I'll try a few (amended) commits and get some additional log output.

**Comment by stianst** [comment]
As the unit tests was breaking our build I merged the fix here:
https://github.com/keycloak/keycloak/commit/f63c60855efd71427f678dc5ce73691937e3a6fb

Travis wasn't running unit tests and that's fixed now so this shouldn't happen again.

File: services/src/main/java/org/keycloak/broker/saml/SAMLEndpoint.java [hunk 1]
```diff
@@ -371,13 +371,12 @@ protected Response handleLoginResponse(String samlResponse, SAMLDocumentHolder h
...
                 }
 
-                if (config.isWantAssertionsSigned() && config.isValidateSignature()) {
-                    if (!AssertionUtil.isSignatureValid(assertionElement, getIDPKeyLocator())) {
-                        logger.error("validation failed");
-                        event.event(EventType.IDENTITY_PROVIDER_RESPONSE);
-                        event.error(Errors.INVALID_SIGNATURE);
-                        return ErrorPage.error(session, Messages.INVALID_REQUESTER);
-                    }
+                if ((config.isWantAssertionsSigned() && !XMLSignatureUtil.isSigned(assertionElement.getOwnerDocument()))
```
Comment (hmlnarik) [inline services/src/main/java/org/keycloak/broker/saml/SAMLEndpoint.java]: WantAssertionsSigned requires signed assertions while what is checked for validity in isSigned is the whole document (arbitrary element signature would suffice to satisfy the condition in that method). This has to be fixed to reflect WantAssertionsSigned logic.

File: saml-core/src/test/resources/org/keycloak/saml/processing/core/parsers/saml/v2/util/saml20-signed-response.xml [hunk 2]
```diff
@@ -0,0 +1 @@
...rmation></saml:Subject><saml:Conditions NotBefore="2017-05-04T22:18:11.546Z" NotOnOrAfter="2017-05-04T22:19:11.546Z"><saml:AudienceRestriction><saml:Audience>http://localhost:8080/auth/realms/saml-broker-authentication-realm</saml:Audience></saml:AudienceRestriction></saml:Conditions><saml:AuthnStatement AuthnInstant="2017-05-04T22:18:13.551Z" SessionIndex="c0cc880f-83d7-461b-b299-9fb3354f598c"><saml:AuthnContext><saml:AuthnContextClassRef>urn:oasis:names:tc:SAML:2.0:ac:classes:unspecified</saml:AuthnContextClassRef></saml:AuthnContext></saml:AuthnStatement><saml:AttributeStatement><saml:Attribute Name="Role" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:basic"><saml:AttributeValue xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:type="xs:string">manager</saml:AttributeValue></saml:Attribute><saml:Attribute Name="Role" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:basic"><saml:AttributeValue xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:type="xs:string">user</saml:AttributeValue></saml:Attribute></saml:AttributeStatement></saml:Assertion></samlp:Response>
```
Comment (hmlnarik) [inline saml-core/src/test/resources/org/keycloak/saml/processing/core/parsers/saml/v2/util/saml20-signed-response.xml]: This XML needs to be properly formatted, but actually see below.

File: saml-core/src/main/java/org/keycloak/saml/processing/core/saml/v2/util/AssertionUtil.java [hunk 3]
```diff
@@ -272,24 +272,21 @@ public static void createSAML11TimedConditions(SAML11AssertionType assertion, lo
...
+     *
+     * @param assertionElement
+     * @param keyLocator the {@link KeyLocator}
+     *
+     * @return true if signature is valid
      */
+    
     public static boolean isSignatureValid(Element assertionElement, KeyLocator keyLocator) {
         try {
             Document doc = DocumentUtil.createDocument();
```
Comment (hmlnarik) [inline saml-core/src/main/java/org/keycloak/saml/processing/core/saml/v2/util/AssertionUtil.java]: Fix to the previous PR that introduced this method - this method should use XMLSignatureUtil.validateSingleNode() (make it public beforehand) instead of reimplementing the logic here.

File: saml-core/src/main/java/org/keycloak/saml/processing/core/saml/v2/util/AssertionUtil.java [hunk 4]
```diff
@@ -557,21 +554,25 @@ public static AssertionType getAssertion(ResponseType responseType, PrivateKey p
         return responseType.getAssertions().get(0).getAssertion();
     }
 
-    public static boolean isAssertionEncrypted(ResponseType responseType) throws ProcessingException {
+    public static boolean isAssertionsEncrypted(ResponseType responseType) throws ProcessingException {
```
Comment (hmlnarik) [inline saml-core/src/main/java/org/keycloak/saml/processing/core/saml/v2/util/AssertionUtil.java]: Currently Keycloak only uses first assertion and this PR does not change this behaviour. Why to introduce this change?

File: saml-core/src/main/java/org/keycloak/saml/processing/core/util/IDFedLSInputResolver.java [hunk 5]
```diff
@@ -113,7 +114,7 @@ public LSInput resolveResource(String type, String namespaceURI, final String pu
         LSInput lsi = null;
         if (systemId == null)
             throw logger.nullValueError("systemid");
-        if (StringUtil.isNotNull(systemId) && systemId.endsWith("dtd") && StringUtil.isNotNull(baseURI)) {
+        if (StringUtil.isNotNull(systemId) && !systemId.endsWith("dtd") && StringUtil.isNotNull(baseURI)) {
```
Comment (hmlnarik) [inline saml-core/src/main/java/org/keycloak/saml/processing/core/util/IDFedLSInputResolver.java]: This change is unrelated to KEYCLOAK-3056 and lacks tests. Remove.

File: saml-core/src/main/java/org/keycloak/saml/processing/core/util/IDFedLSInputResolver.java [hunk 6]
```diff
@@ -49,6 +49,7 @@
         // XML Schema/DTD
         schemaLocationMap.put("datatypes.dtd", "schema/w3c/xmlschema/datatypes.dtd");
         schemaLocationMap.put("XMLSchema.dtd", "schema/w3c/xmlschema/XMLSchema.dtd");
+        schemaLocationMap.put("http://www.w3.org/2001/XMLSchema.dtd", "schema/w3c/xmlschema/XMLSchema.dtd");
```
Comment (hmlnarik) [inline saml-core/src/main/java/org/keycloak/saml/processing/core/util/IDFedLSInputResolver.java]: This change is unrelated to KEYCLOAK-3056 and lacks tests. Remove.

File: services/src/main/java/org/keycloak/broker/saml/SAMLEndpoint.java [hunk 7]
```diff
@@ -351,10 +351,10 @@ protected Response handleLoginResponse(String samlResponse, SAMLDocumentHolder h
                     return callback.error(relayState, Messages.IDENTITY_PROVIDER_UNEXPECTED_ERROR);
                 }
 
-                boolean assertionIsEncrypted = AssertionUtil.isAssertionEncrypted(responseType);
+                boolean assertionIsEncrypted = AssertionUtil.isAssertionsEncrypted(responseType);
```
Comment (hmlnarik) [inline services/src/main/java/org/keycloak/broker/saml/SAMLEndpoint.java]: Currently Keycloak only uses first assertion and this PR does not change this behaviour. Why to introduce this change?

File: services/src/main/java/org/keycloak/broker/saml/SAMLEndpoint.java [hunk 8]
```diff
@@ -351,10 +351,10 @@ protected Response handleLoginResponse(String samlResponse, SAMLDocumentHolder h
...
 
                 if (config.isWantAssertionsEncrypted() && !assertionIsEncrypted) {
-                    logger.error("The assertion is not encrypted, which is required.");
+                    logger.error("The assertions are not encrypted, which is required.");
```
Comment (hmlnarik) [inline services/src/main/java/org/keycloak/broker/saml/SAMLEndpoint.java]: Currently Keycloak only uses first assertion and this PR does not change this behaviour. Why to introduce this change?

File: services/src/main/java/org/keycloak/protocol/saml/SamlService.java [hunk 9]
```diff
@@ -580,7 +580,7 @@ public static String getIDPMetadataDescriptor(UriInfo uriInfo, KeycloakSession s
         return StringPropertyReplacer.replaceProperties(template, props);
     }
 
-    private static void addKeyInfo(StringBuilder target, RsaKeyMetadata key, String purpose) {
+    protected static void addKeyInfo(StringBuilder target, RsaKeyMetadata key, String purpose) {
```
Comment (hmlnarik) [inline services/src/main/java/org/keycloak/protocol/saml/SamlService.java]: What is reason for this change?

**Review by hmlnarik — CHANGES_REQUESTED** [review]
This PR currently touches several areas at once, including those that are not contributing to KEYCLOAK-3056. Please remove unrelated changes. Furthermore please remove tests with mocks.

**Comment by skjolber** [comment]
I was unaware that Keycloak just uses the first assertion, so I'll remove those changes for multiple assertions. Properties name like wantAssertionsEncrypted got me confused.

File: saml-core/src/main/java/org/keycloak/saml/processing/core/util/IDFedLSInputResolver.java (hunk 6 above)
Comment (skjolber) [inline saml-core/src/main/java/org/keycloak/saml/processing/core/util/IDFedLSInputResolver.java]: The previous PR changed the SPMetadataDescriptor which generates XML. So the unit test ValidationTest was updated to actually schema-validate the generated output. However this did not work right without a few fixes to IDFedLSInputResolver. Constructing a dedicated test for the resolver is quite difficult, since it need to interfere with outgoing http connections. The best way is to just disconnect the internet and see that schema loading still works. Any ideas?

File: services/src/main/java/org/keycloak/protocol/saml/SamlService.java (hunk 9 above)
Comment (skjolber) [inline services/src/main/java/org/keycloak/protocol/saml/SamlService.java]: Leftover from refactoring, removing.

**Comment by hmlnarik** [comment]
That's true. Keycloak broker assertion processing only uses first assertion though, see https://github.com/skjolber/keycloak/blob/a074a17ce263b33acdcb2a200ca435f5081d4e26/services/src/main/java/org/keycloak/broker/saml/SAMLEndpoint.java#L382

File: saml-core/src/main/java/org/keycloak/saml/processing/core/util/IDFedLSInputResolver.java (hunk 6 above)
Comment (skjolber) [inline saml-core/src/main/java/org/keycloak/saml/processing/core/util/IDFedLSInputResolver.java]: Added an unit test for IDFedLSInputResolver, was not so hard after all.

**Comment by skjolber** [comment]
Removed mocking. Fixed unit test for IDFedLSInputResolver. Signature checks should now be for a single Assertion (as a direct child).

Was not able to pretty-print saml20-signed-response.xml without breaking signature check.

File: testsuite/integration/src/test/java/org/keycloak/testsuite/saml/ValidationTest.java [hunk 10]
```diff
@@ -16,81 +16,85 @@
  */
```
Comment (hmlnarik) [inline testsuite/integration/src/test/java/org/keycloak/testsuite/saml/ValidationTest.java]: Subject for another JIRA. Remove.

File: saml-core/src/main/java/org/keycloak/saml/common/util/DocumentUtil.java [hunk 11]
```diff
@@ -554,4 +554,33 @@ private static DocumentBuilderFactory getDocumentBuilderFactory() {
...
+        Node child = parent.getFirstChild();
+        
+        while(child != null) {
+            if(child instanceof Element) {
+                Element childElement = (Element)child;
+                
+                String ns = childElement.getNamespaceURI();
+                String localName = childElement.getLocalName();
+                
+                if(targetNamespace.equals(ns) && targetLocalName.equals(localName)) {
```
Comment (hmlnarik) [inline saml-core/src/main/java/org/keycloak/saml/common/util/DocumentUtil.java]: Use ```Objects.equals(targetNamespace, ns)``` (dtto for local name) instead to prevent accidental NPE

File: saml-core/src/main/java/org/keycloak/saml/processing/api/saml/v2/sig/SAML2Signature.java [hunk 12]
```diff
@@ -210,18 +208,20 @@ public Node getNextSiblingOfIssuer(Document doc) {
...
         for (int i = 0; i < nodes.getLength(); i++) {
-            Node n = nodes.item(i);
-            if (n instanceof Element) {
-                ((Element) n).setIdAttribute(ID_ATTRIBUTE_NAME, true);
-            }
+            configureIdAttribute((Element) nodes.item(i));
         }
     }
+    
+    public static void configureIdAttribute(Element assertion) {
```
Comment (hmlnarik) [inline saml-core/src/main/java/org/keycloak/saml/processing/api/saml/v2/sig/SAML2Signature.java]: This method works for any element, not just assertion, please rename parameter accordingly.

**Comment by hmlnarik** [comment]
I see this PR having a good point in fixing the SAMLEndpoint condition and some refactorings. Let's stick to these. Please leave the resolver changes out and create separate JIRA for them to have them fixed properly in a separate PR.

**Review by hmlnarik — CHANGES_REQUESTED** [review]
.

**Comment by skjolber** [comment]
Ok, so lets go with the resolver first, before finalizing this PR. I've created https://issues.jboss.org/browse/KEYCLOAK-4979, but am not sure where to copy the picketlink schema. saml-core is my best suggestion.

File: saml-core/src/main/java/org/keycloak/saml/processing/api/saml/v2/sig/SAML2Signature.java [hunk 13]
```diff
@@ -210,18 +208,20 @@ public Node getNextSiblingOfIssuer(Document doc) {
...
         // Estabilish the IDness of the ID attribute.
-        document.getDocumentElement().setIdAttribute(ID_ATTRIBUTE_NAME, true);
+        configureIdAttribute(document.getDocumentElement());
 
         NodeList nodes = document.getElementsByTagNameNS(JBossSAMLURIConstants.ASSERTION_NSURI.get(),
                 JBossSAMLConstants.ASSERTION.get());
 
         for (int i = 0; i < nodes.getLength(); i++) {
-            Node n = nodes.item(i);
-            if (n instanceof Element) {
```
Comment (skjolber) [inline saml-core/src/main/java/org/keycloak/saml/processing/api/saml/v2/sig/SAML2Signature.java]: document.getElementsByTagNameNS(...) returns only Element, so the if statement is always true.

**Comment by hmlnarik** [comment]
I appreciate your will to fix the resolver but that is of lower priority than this PR. Please fix this PR as requested.

**Comment by skjolber** [comment]
This PR has been updated as requested.

File: services/src/main/java/org/keycloak/broker/saml/SAMLEndpoint.java [hunk 14]
```diff
@@ -371,13 +371,12 @@ protected Response handleLoginResponse(String samlResponse, SAMLDocumentHolder h
...
-                if (config.isWantAssertionsSigned() && config.isValidateSignature()) {
-                    if (!AssertionUtil.isSignatureValid(assertionElement, getIDPKeyLocator())) {
-                        logger.error("validation failed");
-                        event.event(EventType.IDENTITY_PROVIDER_RESPONSE);
-                        event.error(Errors.INVALID_SIGNATURE);
-                        return ErrorPage.error(session, Messages.INVALID_REQUESTER);
-                    }
+                if ((config.isWantAssertionsSigned() && !AssertionUtil.isSignedElement(assertionElement))
+                        || (config.isValidateSignature() && !AssertionUtil.isSignatureValid(assertionElement, getIDPKeyLocator()))) {
+                    logger.error("validation failed");
```
Comment (hmlnarik) [inline services/src/main/java/org/keycloak/broker/saml/SAMLEndpoint.java]: Would this condition work for SAML documents where only the whole document (not the assertions) is signed?
Please create a KcSamlSignedDocumentOnlyBrokerTest based on org.keycloak.testsuite.broker.KcSamlSignedBrokerTest, check https://github.com/keycloak/keycloak/blob/master/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/KcSamlSignedBrokerTest.java#L52.

File: saml-core/src/main/java/org/keycloak/saml/processing/api/saml/v2/sig/SAML2Signature.java (hunk 13 above)
Comment (hmlnarik) [inline saml-core/src/main/java/org/keycloak/saml/processing/api/saml/v2/sig/SAML2Signature.java]: That's true according to the documentation, but in reality it returns an untyped NodeList where by mistake a non-Element node can occur. Better safe than sorry here, please keep the condition.

File: saml-core/src/main/java/org/keycloak/saml/processing/core/util/XMLSignatureUtil.java [hunk 15]
```diff
@@ -409,7 +409,7 @@ public static Document sign(SignatureUtilTransferObject dto, String canonicaliza
 
         return doc;
     }
-
+
```
Comment (hmlnarik) [inline saml-core/src/main/java/org/keycloak/saml/processing/core/util/XMLSignatureUtil.java]: Please remove this change.

**Review by hmlnarik — CHANGES_REQUESTED** [review]
Thank you. Please see comments in individual lines

File: services/src/main/java/org/keycloak/broker/saml/SAMLEndpoint.java (hunk 14 above)
Comment (skjolber) [inline services/src/main/java/org/keycloak/broker/saml/SAMLEndpoint.java]: Does signing the whole document cound as signing the assertion?

File: saml-core/src/main/java/org/keycloak/saml/processing/api/saml/v2/sig/SAML2Signature.java (hunk 13 above)
Comment (skjolber) [inline saml-core/src/main/java/org/keycloak/saml/processing/api/saml/v2/sig/SAML2Signature.java]: If the Document implementation is not according to spec, and the cast to Element fails, we then would be aware of and able to fix the bug in the implementation. If we do not trust the implementation to work, why bother calling the method - it might 'not work' in countless ways. The code should be safe, as there is no logical loophole (an exception will always occor). 

So unless there is a concrete example of an implemenation which triggers an exception in the code, on which the project is dependant and which cannot be fixed, keeping the old code does not make sense.

File: services/src/main/java/org/keycloak/broker/saml/SAMLEndpoint.java (hunk 14 above)
Comment (hmlnarik) [inline services/src/main/java/org/keycloak/broker/saml/SAMLEndpoint.java]: I don't understand your question. The changed code checks for validity of Assertion element regardless of the wantAssertionsSigned flag. The IdP can be configured to sign the whole document only, not individual assertions. Is this case still working? The PR has to contain an integration test for that.

Comment (skjolber) [inline services/src/main/java/org/keycloak/broker/saml/SAMLEndpoint.java]: Okey, I understand better now.

**Comment by hmlnarik** [comment]
@skjolber When do you expect to update this PR?

File: services/src/main/java/org/keycloak/broker/saml/SAMLEndpoint.java (hunk 14 above)
Comment (skjolber) [inline services/src/main/java/org/keycloak/broker/saml/SAMLEndpoint.java]: If validateSignature is true, should the signature be check regardless of whether the assertion was wanted or not?

**Comment by skjolber** [comment]
@hmlnarik Made some adjustments now, but frankly not sure the KcSamlSignedDocumentOnlyBrokerTest is ready.

**Comment by hmlnarik** [comment]
> Made some adjustments now, but frankly not sure the KcSamlSignedDocumentOnlyBrokerTest is ready.
Why? It seems correct to me.

> Edit: So the validateSignature setting, does it imply that Assertion signatures should be checked, regardless of whether they were expected (wanted) or not?
Yes.

**Review by hmlnarik — APPROVED** [review]
@skjolber Thank you
@stianst This PR can be merged.

**Comment by skjolber** [comment]
@hmlnarik Thanks for reviewing!
